### PR TITLE
Fix #60: Crash on Android when onCompleted is called after Fetch instance is closed

### LINKED ISSUE #60: Crash on Android when app is exited mid-download
### Is this a bug report, a feature request, or a question?

Bug report.

### Have you followed the required steps before opening a bug report?

- [X] I have reviewed [the documentation](https://github.com/EkoLabs/react-native-background-downloader/blob/master/README.md) in its entirety, including the dedicated documentations :books:.
- [X] I have searched for [existing issues](https://github.com/EkoLabs/react-native-background-downloader/issues) and made sure that the problem hasn't already been reported.
- [X] I am using [the latest plugin version](https://github.com/EkoLabs/react-native-background-downloader/releases).

### Is the bug specific to iOS or Android? Or can it be reproduced on both platforms?

Android.

### Is the bug related to the native implementation? (NSURLSession on iOS and Fetch on Android)

Yes, Fetch on Android. The onCompleted callback implemented by this project contains an assumption that the fetch instance is available when it is called (which apparently is not always true).

### Environment

  Environment:
    react-native-background-downloader: 2.3.1

  Target Platform:
    Android (6.0) and Android (10.0)

### Expected Behavior

No exception is thrown during the exit of the app.

### Actual Behavior

When the app is closed an exception is thrown:
`com.tonyodev.fetch2.fetch.FetchImpl.throwExceptionIfClosed FetchImpl.kt:1226
com.tonyodev.fetch2.fetch.FetchImpl.executeRemoveAction FetchImpl.kt:440
com.tonyodev.fetch2.fetch.FetchImpl.remove FetchImpl.kt:385
com.tonyodev.fetch2.fetch.FetchImpl.remove FetchImpl.kt:397
com.tonyodev.fetch2.fetch.FetchImpl.remove FetchImpl.kt:393
com.eko.RNBackgroundDownloaderModule.onCompleted RNBackgroundDownloaderModule.java:315
com.tonyodev.fetch2.fetch.ListenerCoordinator$mainListener$1$onCompleted$$inlined$synchronized$lambda$2.run ListenerCoordinator.kt:246
android.os.Handler.handleCallback Handler.java:873
android.os.Handler.dispatchMessage Handler.java:99
android.os.Looper.loop Looper.java:216
android.app.ActivityThread.main ActivityThread.java:7266
java.lang.reflect.Method.invoke Method.java
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run RuntimeInit.java:494
com.android.internal.os.ZygoteInit.main ZygoteInit.java:975`

### Steps to Reproduce

This issue does not always trigger. It is related to when the fetch instance is closed and when onCompleted callbacks are executed. I was able to reproduce it twice by exiting the Android app using this lib mid-download.

### PR DESCRIPTION
**Pull request by mark-plukkido**
This PR fixes #60 which is a crash which happens when download is interrupted by exiting the app on Android and the Fetch instance is already closed by the time the onComplete callback is called.

### CODE REVIEW

File: ios/RNBackgroundDownloader.m [hunk 1]
```diff
@@ -79,17 +79,21 @@ - (void)lazyInitSession {
...
-    [taskToConfigMap removeObjectForKey:taskId];
-    [[NSUserDefaults standardUserDefaults] setObject:[self serialize: taskToConfigMap] forKey:ID_TO_CONFIG_MAP_KEY];
-    if (taskConfig) {
-        [idToTaskMap removeObjectForKey:taskConfig.id];
-        [idToPercentMap removeObjectForKey:taskConfig.id];
-    }
-    if (taskToConfigMap.count == 0) {
-        [urlSession invalidateAndCancel];
-        urlSession = nil;
+    @synchronized (sharedLock) {
```
Comment (behealy) [inline ios/RNBackgroundDownloader.m]: @mark-plukkido Just curious, was it your intention to reverse this commit: https://github.com/EkoLabs/react-native-background-downloader/pull/61/commits/44eaa8c87130bd5c68a37e9c107773e42a443d07?

Comment (mark-plukkido) [reply ios/RNBackgroundDownloader.m]: No, absolutely not. I made a mistake when I merged from upstream and then in 5d86a50 I changed it back to your code. There should be only one change in RNBackgroundDownloader.java. 

Do you want me to create a new PR without these? In that case, just reject this PR and I recreate it without the extra confusion.

Thanks!

Comment (behealy) [reply ios/RNBackgroundDownloader.m]: @ptelad Could you reject this PR so @mark-plukkido Can recreate it? I don't think I have the ability to reject as I am not a contributor/owner in this repo. I was just looking at the latest changes and PRs and I noticed something was off with this one.

Comment (behealy) [reply ios/RNBackgroundDownloader.m]: @mark-plukkido Oh I see what happened. I was just looking through these new PRs and got a little confused. Its probably up to you and @ptelad whether or not you'd like to remake the PR.

**Comment by ptelad** [comment]
merged and published in `2.3.2`

**Pull request merged**